src/core/qgsmaprenderertask.h

```diff
@@ -0,0 +1,97 @@
+/***************************************************************************
+                          qgsmaprenderertask.h
+                          -------------------------
+    begin                : Apr 2017
+    copyright            : (C) 2017 by Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMAPRENDERERTASK_H
+#define QGSMAPRENDERERTASK_H
+
+#include "qgis_core.h"
+#include "qgsannotation.h"
+#include "qgsannotationmanager.h"
+#include "qgsmapsettings.h"
+#include "qgstaskmanager.h"
+
+#include <QPainter>
+
+/**
+ * \class QgsMapRendererTask
+ * \ingroup core
+ * QgsTask task which draws a map to an image file or a painter as a background
+ * task. This can be used to draw maps without blocking the QGIS interface.
+ * \since QGIS 3.0
+ */
+class CORE_EXPORT QgsMapRendererTask : public QgsTask
+{
+    Q_OBJECT
+
+  public:
+
+    enum ErrorType
+    {
+      ImageAllocationFail = 1,
+      ImageSaveFail
+    };
+
+    /**
+     * Constructor for QgsMapRendererTask to render a map to an image file.
+     */
+    QgsMapRendererTask( const QgsMapSettings &ms,
+                        const QString &fileName,
+                        const QString &fileFormat = QStringLiteral( "PNG" ) );
```

Comment:
I've had issues with QStringLiteral use in definition arguments before - better to change this to a standard QString("PNG")